library/index.c

```diff
@@ -323,24 +323,25 @@ int eblob_index_blocks_fill(struct eblob_base_ctl *bctl)
 }
 
 
-static struct eblob_disk_control *eblob_find_on_disk(struct eblob_backend *b,
-		struct eblob_base_ctl *bctl, struct eblob_disk_control *dc,
+static int eblob_find_on_disk(struct eblob_backend *b,
```

Comment:
bsearch has to run over single block (4k in size) only. There is a range index tree which contains ranges of keys and their start/stop positions in index. This range tree is generated on eblob start. When we have to find a key, we find a range containing this key and perform binary search in it. Ranges do not overlap, but can contain multiple key entries because of object removal for example.

You should not worry about random read pattern - linux kernel has readahead subsystem which analyzes read pattern and reads ahead part of the file into page cache. Also, index files are frequently accessed and do not flushed out of page cache.

Similar logic should be implemented with pread() approach - read not the whole file but only a chunk containing range, which wraps over they key we are interested in, and then use part of the old eblob_find_on_disk() which ran bsearch(). You can only replace search_start/search_end setting - not from mapped area, but reading it from file descriptor.